### ISSUE TITLE
Hi, about keras of dtln-biLSTM？

### ISSUE THREAD

**Issue by ghost** [eagomez2/upf-smc-speech-enhancement-thesis#1]
keras：
Layer (type)                    Output Shape         Param #     Connected to                     
==================================================================================================
input_1 (InputLayer)            [(None, None)]       0                                            
__________________________________________________________________________________________________
lambda (Lambda)                 [(None, None, 257),  0           input_1[0][0]                    
__________________________________________________________________________________________________
tf_op_layer_AddV2 (TensorFlowOp [(None, None, 257)]  0           lambda[0][0]                     
__________________________________________________________________________________________________
tf_op_layer_Log (TensorFlowOpLa [(None, None, 257)]  0           tf_op_layer_AddV2[0][0]          
__________________________________________________________________________________________________
instant_layer_normalization (In (None, None, 257)    514         tf_op_layer_Log[0][0]            
__________________________________________________________________________________________________
bidirectional (Bidirectional)   (None, None, 256)    395264      instant_layer_normalization[0][0]
__________________________________________________________________________________________________
dropout (Dropout)               (None, None, 256)    0           bidirectional[0][0]              
__________________________________________________________________________________________________
bidirectional_1 (Bidirectional) (None, None, 256)    394240      dropout[0][0]                    
__________________________________________________________________________________________________
dense (Dense)                   (None, None, 257)    66049       bidirectional_1[0][0]            
__________________________________________________________________________________________________
activation (Activation)         (None, None, 257)    0           dense[0][0]                      
__________________________________________________________________________________________________
multiply (Multiply)             (None, None, 257)    0           lambda[0][0]                     
                                                                 activation[0][0]                 
__________________________________________________________________________________________________
lambda_1 (Lambda)               (None, None, 512)    0           multiply[0][0]                   
                                                                 lambda[0][1]                     
__________________________________________________________________________________________________
conv1d (Conv1D)                 (None, None, 256)    131072      lambda_1[0][0]                   
__________________________________________________________________________________________________
instant_layer_normalization_1 ( (None, None, 256)    512         conv1d[0][0]                     
__________________________________________________________________________________________________
bidirectional_2 (Bidirectional) (None, None, 256)    394240      instant_layer_normalization_1[0][
__________________________________________________________________________________________________
dropout_1 (Dropout)             (None, None, 256)    0           bidirectional_2[0][0]            
__________________________________________________________________________________________________
bidirectional_3 (Bidirectional) (None, None, 256)    394240      dropout_1[0][0]                  
__________________________________________________________________________________________________
dense_1 (Dense)                 (None, None, 256)    65792       bidirectional_3[0][0]            
__________________________________________________________________________________________________
activation_1 (Activation)       (None, None, 256)    0           dense_1[0][0]                    
__________________________________________________________________________________________________
multiply_1 (Multiply)           (None, None, 256)    0           conv1d[0][0]                     
                                                                 activation_1[0][0]               
__________________________________________________________________________________________________
conv1d_1 (Conv1D)               (None, None, 512)    131072      multiply_1[0][0]                 
__________________________________________________________________________________________________
lambda_2 (Lambda)               (None, None)         0           conv1d_1[0][0]                   
==================================================================================================
Total params: 1,972,995
Trainable params: 1,972,995
Non-trainable params: 0

Your torch: 
DTLN summary:
+----+----------------------------------------------+--------------------+------------+
| #  |                    Layer                     |        Type        | Parameters |
+----+----------------------------------------------+--------------------+------------+
| 0  |                 layer_norm_0                 |     LayerNorm      |    514     |
| 1  |              separation_core_0               | DTLNSeparationCore |   857601   |
| 2  |      separation_core_0.separation_core       |     ModuleDict     |   857601   |
| 3  | separation_core_0.separation_core.rnn_stack  |        LSTM        |   791552   |
| 4  |    separation_core_0.separation_core.fcl     |       Linear       |   66049    |
| 5  | separation_core_0.separation_core.activation |      Sigmoid       |     0      |
| 6  |                   conv1d_1                   |       Conv1d       |   131072   |
| 7  |                 layer_norm_1                 |     LayerNorm      |    512     |
| 8  |              separation_core_1               | DTLNSeparationCore |   856320   |
| 9  |      separation_core_1.separation_core       |     ModuleDict     |   856320   |
| 10 | separation_core_1.separation_core.rnn_stack  |        LSTM        |   790528   |
| 11 |    separation_core_1.separation_core.fcl     |       Linear       |   65792    |
| 12 | separation_core_1.separation_core.activation |      Sigmoid       |     0      |
| 13 |               causal_conv1d_1                |    CausalConv1d    |   131072   |
| 14 |               overlap_and_add                |        Fold        |     0      |
+----+----------------------------------------------+--------------------+------------+
Total parameters: 1977091
Trainable parameters: 1977091
Non-trainable parameters: 0
Hi，do you know what is the difference between keras and torch ？

**Comment by eagomez2** [comment]
Hi zqlsnr,

Keras and PyTorch are two different libraries to work with neural networks.  I decided to work with PyTorch because I felt more comfortable working with it in a research oriented manner as compared to the subclassing API from tensorflow. In theory, if you change the LSTM layers of DTLN for BiLSTM, you should obtain statistically similar results to my PyTorch translation of [the repository by Nils et. al](https://github.com/breizhn/DTLN) set to `-m dtln_bilstm`. 
That being said, some components considered important in deep learning such as weight initialization or the underlying low level implementation of different layers may be different, and could have an impact on the overall result. Moreover, some people may argue that using the same architecture they obtain different results just by changing libraries:

https://discuss.pytorch.org/t/pytorch-vs-tensorflow-gives-different-results/4630
https://stackoverflow.com/questions/66039582/image-classification-why-am-i-getting-vastly-different-results-training-on-tens
https://github.com/tensorflow/tensorflow/issues/7624

So far I haven't seen people being able to successfully isolate what would cause this difference for a given case, that would not be an easy task, specially considering the black box nature of many deep learning applications.

All the best,
Esteban

**Comment by ghost** [comment]
Hi, eagomez2 
The reason may be that the bias is different. In pytorch's LSTM, the definition of bias is split into two parts, so each Torch's LSTM is 1024 more than Keras' LSTM...
![image](https://user-images.githubusercontent.com/83391979/131959029-75968e8a-11af-4a4d-8cfd-16c239295077.png)

**Comment by eagomez2** [comment]
Hi zqlsnr,

If you found that difference, it may definitely contribute to the implementation differences I mentioned earlier. In the case of PyTorch the `ih_l0` is the input-hidden bias and `hh_l0` is the hidden-hidden bias as explained here:
https://pytorch.org/docs/stable/generated/torch.nn.LSTM.html

Keras LSTM has also a `recurrent_initializer`, regularization options and a bunch of others that are not available off-the-shelf in PyTorch, as well as the `stateful` option that you have to manually implement in the case of PyTorch:
https://www.tensorflow.org/api_docs/python/tf/keras/layers/LSTM

Other differences could also be found in different types of layer that will add up to make a similar implementation work differently, as well as low level details such as the order of operations and so on that may have some effect on the obtained result rounding error.

**Comment by ghost** [comment]
Hi eagomez2,
Thank you! I will try to train it according to your suggestion. I am currently training a 48000 sample rate model. Is the API for real-time processing of CRUSE and dtln-bilstm implemented through sounddevice？

**Comment by eagomez2** [comment]
Hi zqlsnr,
 
 There is currently no real time implementation of these models in the repository, although I would love to add one similarly [denoiser](https://github.com/facebookresearch/denoiser). I am thinking of doing so as soon as I make some progress on other projects, but if you already have it or want to contribute, it would be highly appreciated.
